app/internal/statistics.py

```diff
@@ -0,0 +1,391 @@
+        end: end of date range.
+    """
+    if not does_user_exist(session=db, user_id=userid):
+        return ValidateResult(
+            valid_input=False, error_text=INVALID_USER, start=start, end=end)
+    date = start or datetime.datetime.now()
+    start = datetime.datetime(date.year, date.month, date.day)
+    single_day = datetime.timedelta(days=1, seconds=-1)
+    date = end or start
+    end = datetime.datetime(date.year, date.month, date.day) + single_day
+    if start >= end:
+        return ValidateResult(
+            valid_input=False, error_text=INVALID_DATE_RANGE,
+            start=start, end=end)
+    return ValidateResult(
+        valid_input=True, error_text="", start=start, end=end)
+
+
+def event_start_end_filter():
+    """
+    prepare generic part of SQL to be used in all queries
+    Returns:
+        concatenated sql part for events date
+    """
+    return """(events.start >= start and events.start <= end)
+            or (events.end >= start and events.end <= end)
+            or (events.start <= start and events.end >= end)
+    """
+
+
+def get_events_count_stats(
+        db: Session, userid: int, start: datetime.datetime,
+        end: datetime.datetime) -> Dict[str, Dict[str, int]]:
+    """ calculate statistics and events relevant for the requested user
+        and the requested date range.
+        all logic is performed in the db.
+
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+
+    Returns: json with
+        events_count_stats: (for requested date range).
+            meetings_for_user: events that the user has.
+            created_by_user: events that the user has created.
+    """
+    meetings_for_user = db.query(Event.id).join(
+        UserEvent, UserEvent.event_id == Event.id).filter(
+        UserEvent.user_id == userid).filter(
+        text(event_start_end_filter())).count()
+    created_by_user = db.query(Event.id).filter(
+        Event.owner_id == userid).filter(
+        text(event_start_end_filter())).count()
+    return {"events_count_stats": {"meetings_for_user": meetings_for_user,
+                                   "created_by_user": created_by_user}}
+
+
+def get_events_by_date(
+        db: Session, userid: int, start: datetime.datetime,
+        end: datetime.datetime) -> List[Tuple[datetime.datetime, int]]:
+    """ get data of date + number of events on it from the db
+
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+
+    Returns:
+        data of date + number of events on it
+    """
+    events_by_date = db.query(
+        func.date(Event.start), func.count(func.date(Event.start))).join(
+        UserEvent, UserEvent.event_id == Event.id).filter(
+        UserEvent.user_id == userid).filter(
+        text(event_start_end_filter())).all()
```

Comment:
We always prefer to split the code in a fashion that will allow others to easily skim through it.

Compare with:
```python
def get_date_filter_between_dates(start: datetime.datetime, end: datetime.datetime) -> ?:
    """
    Prepare the filter by dates using declarative SQLAlchemy.

    Returns:
        concatenated sql part for events date
    """
    event_start, event_end = func.date(Event.start), func.date(Event.end)
    return (
        (Event.start >= start and Event.start <= end)
        or (Event.end >= start and Event.end <= end)
        or (Event.start <= start and Event.end >= end)
    )

def get_events_by_date(
        db: Session, userid: int, start: datetime.datetime,
        end: datetime.datetime) -> List[Tuple[datetime.datetime, int]]:
    """Get (what?) about the date, and number of events on it.

    Args:
        db: db session.
        userid: requested user id number for statistics.
        start: start of date range.
        end: end of date range.

    Returns:
        Number of events on the date range.
    """
    start_date = func.date(Event.start)
    events_count = func.count(start_date)
    by_user_id = UserEvent.user_id == userid
    user_to_event = (UserEvent, UserEvent.event_id == Event.id)
    return (
        db.query(start_date, events_count).join(user_to_event)
        .filter(by_user_id)
        .filter(get_date_filter_between_dates(start, end))
        .all()
    )